### PR TITLE
maturin: adopt; refactor; modernize

### PR DESCRIPTION
**Pull request by getchoo**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ma/maturin/pyo3-test/generic.nix [hunk 1]
```diff
@@ -24,22 +25,25 @@ python.pkgs.buildPythonPackage rec {
...
 
   pythonImportsCheck = [ "word_count" ];
 
-  meta = with lib; {
+  meta = {
     description = "PyO3 word count example";
     homepage = "https://github.com/PyO3/pyo3";
-    license = licenses.asl20;
+    license = lib.licenses.asl20;
     maintainers = [ ];
```
Comment (SuperSandro2000) [inline pkgs/by-name/ma/maturin/pyo3-test/generic.nix:47]: Maybe this, too?

**Review by SuperSandro2000 — COMMENTED** [review]
This PR rebuilds a lot of packages which means we must target staging. Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#rebasing-between-branches-ie-from-master-to-staging) to not potentially ping a lot of people.

**Comment by getchoo** [comment]
@SuperSandro2000 https://github.com/NixOS/nixpkgs/compare/c90a70ac43d97c8a313fd23412f2e7767268546a..afbcad747272742c0a06984cd52dbf395fb46d39 should have fixed the mass rebuild on darwin. i'll wait on ofborg, but this should be good for master still

**Comment by trofi** [comment]
Bisec t says 7dd53c5e66979d1ca16333c8f987c5b3c19e094b `maturin: migrate to by-name` broke the `python3Packages.setuptools-rust.tests` eval as:

```
$ nix build --no-link -f. python3Packages.setuptools-rust.tests
error:
       … while evaluating the attribute 'pyo3'
         at pkgs/development/python-modules/setuptools-rust/default.nix:41:3:
           40|
           41|   passthru.tests.pyo3 = callPackage ./pyo3-test { };
             |   ^
           42|

       … while calling a functor (an attribute set with a '__functor' attribute)
         at lib/customisation.nix:264:13:
          263|     in if missingArgs == {}
          264|        then makeOverridable f allArgs
             |             ^
          265|        # This needs to be an abort so it can't be caught with `builtins.tryEval`,

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: path 'pkgs/development/tools/rust/maturin/pyo3-test/generic.nix' does not exist
```

**Pull request merged**